File: leetcode/remove_nth_node_from_end_of_list/remove_nth_node_from_end_of_list.cpp
/*
 * Time Complexity: O(L)
 *   - L is the length of the linked list.
 *
 * Space Complexity: O(1)
 *   - Only a constant amount of extra space is used (dummy node and pointers).
 */

#include <queue>

using std::queue;

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        if(head == nullptr) return head;

        ListNode* dummy = new ListNode(0, head);
        ListNode* prev = dummy;
        ListNode* curr = head;

        for(int i = 0; i < n; i++) {
            curr = curr->next;
        }

        while(curr != nullptr) {
            prev = prev->next;
            curr = curr->next;
        }

        ListNode* todel = prev->next;
        prev->next = todel->next;
        delete todel;
        return dummy->next;
    }
};

int main() {
    ListNode* head = new ListNode(1, new ListNode(2));
    int n = 2;
    Solution sol;
    head = sol.removeNthFromEnd(head, n);
}

File: leetcode/remove_nth_node_from_end_of_list/remove_nth_node_from_end_of_list.java
package remove_nth_node_from_end_of_list;

class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}

class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        if (head == null) return head;

        ListNode dummy = new ListNode(0, head);
        ListNode prev = dummy;
        ListNode curr = head;

        for (int i = 0; i < n; i++) {
            curr = curr.next;
        }

        while (curr != null) {
            prev = prev.next;
            curr = curr.next;
        }

        ListNode toDelete = prev.next;
        prev.next = toDelete.next;

        return dummy.next;
    }
}
